packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c

```diff
@@ -1342,6 +1369,111 @@ static ssize_t set_qsfp_lpmode(struct device *dev, struct device_attribute *da,
     return status;
 }
 
+static char *
+__strtok_r(char *s, const char *delim, char **last)
+{
+	char *spanp, *tok;
+	int c, sc;
+
+	if (s == NULL && (s = *last) == NULL)
+		return (NULL);
+
+	/*
+	 * Skip (span) leading delimiters (s += strspn(s, delim), sort of).
+	 */
+cont:
+	c = *s++;
+	for (spanp = (char *)delim; (sc = *spanp++) != 0;) {
+		if (c == sc)
+			goto cont;
+	}
+
+	if (c == 0) {		/* no non-delimiter characters */
+		*last = NULL;
+		return (NULL);
+	}
+	tok = s - 1;
+
+	/*
+	 * Scan token (scan for delimiters: s += strcspn(s, delim), sort of).
+	 * Note that delim must have one NUL; we stop if we see that, too.
+	 */
+	for (;;) {
+		c = *s++;
+		spanp = (char *)delim;
+		do {
+			if ((sc = *spanp++) == c) {
+				if (c == 0)
+					s = NULL;
+				else
+					s[-1] = '\0';
+				*last = s;
+				return (tok);
+			}
+		} while (sc != 0);
+	}
+	/* NOTREACHED */
+}
+
+/* Command Format:
+   <Port> <Slave Address> <Data Count> <Register 1> <Data 1 High> <Data 1 Low> <Register 2> <Data 2 High> <Data 2 Low> ...
+ */
+static ssize_t set_phy(struct device *dev, struct device_attribute *da,
+			const char *buf, size_t count)
+{
+    int i = 0;
+	int status;
+    char *token = NULL, *last = NULL, *delim = " ";
+    struct sfp_phy_write_data wdata;
+    char command[NUM_OF_PHY_REGISTERS * 3 + 3] = {0};
+    char *ipmi_phy_tx_data = (char *)&wdata;
+
+    memset(&wdata, 0, sizeof(wdata));
+    memcpy(command, buf, sizeof(command));
```

Comment:
Does this run the risk of a fault if "buf" is near the end of a page, there is no adjacent page, and (buf + sizeof(command)) runs past the end of the page?  Even if there is no fault, should you be copying arbitrary kernel data into your command buffer here?  Would it not be better to take the minimum of sizeof(command) and count?  If count exceeds sizeof(command), should that be reflected in an error return?

Reply:
There is a logic error here.
I should declare a string buffer "char *command "then copy from the "const char *buf" passed into set_phy()

I modified as below
```suggestion
    char *command = NULL;
    char *ipmi_phy_tx_data = (char *)&wdata;

    command = kzalloc(count+1, GFP_KERNEL);
    if (!command) {
        return -ENOMEM;
    }

    memset(&wdata, 0, sizeof(wdata));
    memcpy(command, buf, count);
```

The variable "const char *buf" is guaranteed to be safe to access with the length "count" by sysfs core, the buf is in kernel space for driver to access, the copy_from_user is also hidden in core.
Here is the code slice in file.c (/fs/sysfs/file.c)
```suggestion
static int 
fill_write_buffer(struct sysfs_buffer * buffer, const char __user * buf, size_t count)
{
	int error;

	if (!buffer->page)
		buffer->page = (char *)get_zeroed_page(GFP_KERNEL);
	if (!buffer->page)
		return -ENOMEM;

	if (count >= PAGE_SIZE)
		count = PAGE_SIZE - 1;
	error = copy_from_user(buffer->page,buf,count);
```

So the memcpy from buf is safe.

Please review my changes again.